docs/allocations.md

```diff
@@ -0,0 +1,280 @@
+
+### Collecting the raw data
+
+To get started, let's have `perf` run the program for us and collect the information using the `libc_probe:malloc` we set up above.
+
+```
+perf record --call-graph dwarf,16384 \
+     -m 50000 \
+     -e probe_libc:malloc -- \
+     .build/release/your-program-name
+```
+
+Let's break down this command a little:
+
+- `perf record` instructs `perf` to `record` data, makes sense.
+- `--call-graph dwarf,16384` instructs `perf` to use the [DWARF](http://www.dwarfstd.org) information to create the call graphs. It also sets the maximum stack dump size to 16k which should be enough to get you full stack traces. Unfortunately, using DWARF is rather slow (see below) but it creates the best call graphs for you.
+- `-m 50000`: The size of the ring buffer that `perf` uses to buffer. This is given in multiples of `PAGE_SIZE` (usually 4kB) and especially with DWARF this needs to be pretty huge to prevent data loss.
+- `-e probe_libc:malloc`: Record when the `malloc` probe fires
+
+What you want to see if output like this
+
+```
+<your program's output>
+[ perf record: Woken up 2 times to write data ]
+[ perf record: Captured and wrote 401.088 MB perf.data (49640 samples) ]
+```
+
+What you don't want to see is
+
+```
+[ perf record: Woken up 189 times to write data ]
+Warning:
+Processed 4346 events and lost 144 chunks!
+
+Check IO/CPU overload!
+
+[ perf record: Captured and wrote 30.868 MB perf.data (3817 samples) ]
+```
+
+When `perf` tells you that it lost a number of chunks it means that it lost data. If `perf` lost data, you have a few options:
+
+- Reduce the amount of work your program is doing. For every allocation, `perf` will need to record a stack trace.
+- Reduce the maximum "stack dump" that `perf` records by changing the `--call-graph dwarf` parameter to for example `--call-graph dwarf,2048`. The default is to record a maximum of 4096 bytes which gives you pretty deep stacks, if you don't need that you can reduce the number. The tradeoff is that the flame graph may show you `[unknown]` stack frames which means that there are missing stack frames there. The unit is bytes.
+- You can raise the number of the `-m` parameter which is the size of the ring buffer that `perf` uses in memory (in multiples of `PAGE_SIZE`, usually that is 4kB)
+- You can give up nice call graphs and replace `--call-tree dwarf` with `--call-tree fp` (`fp` stands for frame pointer).
+
+### Flame graphs
+
+After a successful `perf record`, you can invoke the following command line to produce an SVG file with the flame graph
+
+```bash
+perf script | \
+    /FlameGraph/stackcollapse-perf.pl - | \
+    swift demangle --simplified | \
+    /FlameGraph/flamegraph.pl --countname allocations \
+        --width 2000 > out.svg
+```
+
+Let's expand a little on what the above command does:
+
+- It runs `perf script` which dumps the binary information that `perf record` recorded into a textual form.
```

Comment:
I went back and forth on that :). I left it out in the this version now because of these reasons:

- even more noise in the cmdline
- it's easy to accidentally `-i` an _old_ `perf.data.TIMESTAMP`
- progressive disclosure

I'm happy to change it again but I wasn't sure if this is necessary here. Tbh that's something I struggled with _a lot_: What to include and what to strip. Other examples:

- default stack shot size vs. 16384 or something else big
- flame _charts_ yes/no, I would've said yes but `stackcollapse-perf.pl` currently can't make them (have a patch)
- `perf` multi-output mode where you switch the output on a certain condition (like a signal or an event)
- the `| sed` stuff to simplify some stack traces (which I included)
- other stuff I now forgot

The truth is that `perf` is super powerful but it has a learning curve that I think is easy enough to overcome if you know what you're looking for but very hard to overcome if you don't. Terminology in the man pages etc and the extra complexity of a non-C language obviously add to that.

I'm very open to suggestions.

Reply:
Oh I don't think it should be in all the examples here, but it's useful to know about, how about a note in "Other `perf` tricks"?

Reply:
Flame charts would be useful too, I think.